### DISCUSSION TITLE
ignore: Matching using overrides and ignore files

### DISCUSSION THREAD

**Discussion by JohnnyMorganz** [BurntSushi/ripgrep#1783] [Q&A] [answered]
Hey 👋 

I've been recently using the `ignore` crate, and its been serving pretty much all my needs (custom ignore files, ignore globs, walk builder to iterate through these files), first of all, thanks for it!

One thing that a user of mine reported was that globs inside of a `.styluaignore` (a custom ignore file I added using `walk_builder.add_custom_ignore_filename`), weren't being matched. Users can currently pass globs as a command line argument which are passed to an OverrideBuilder, and as a default, I automatically add in a `**/*.lua` glob if none were specified (This is a Lua code formatter, so I want it to format Lua files only!)

From what I can tell, because of this glob I'm adding as an override by default, if a user then adds a specific glob to their ignore file to ignore, it won't be used (e.g. a user added `**/*.spec.lua` and `vendor/` to their ignore file - the latter was ignored, but the former wasn't). 

Having further read the docs, it says
> First, glob overrides are checked. If a path matches a glob override, then matching stops. The path is then only skipped if the glob that matched the path is an ignore glob. (An override glob is a whitelist glob unless it starts with a !, in which case it is an ignore glob.)

So I'm assuming its failing at this case - no further tests are being done to see whether to ignore the file. I was wondering, is there any way to go about what I want? Ideally there would be a way to toggle whether this stops matching completely (as current), or whether the whitelist glob acts as a starting thing to check and then continues onwards if matched (what I think would work for me).
Any suggestions would be appreciated, thanks!

**Comment by BurntSushi** [comment, marked as answer]
No, there is no way to change this behavior. That's the purpose of overrides: to override everything else.

If you only want to look at Lua files, then just do that. Ignore anything that isn't a Lua file when visiting each directory entry.

**Reply by JohnnyMorganz** [reply]
Thanks for the quick response!

My use case here was that the user could optionally provide the globs to match, and if they provided nothing, I would fall back to my original default - so I can't just automatically test against the `**/*.lua` glob in case the user *doesn't* want to match against that (e.g. a user wanted to only format `**/*.client.lua` files, or was using it on a codebase which doesn't use `.lua` as the file extension ending). Unless I'm understanding you wrong, I don't think I could do what you suggested.

**Reply by BurntSushi** [reply]
Yes, I don't see an easy way to do what you want, sorry. I suppose what you could do is [add a global ignore file](https://docs.rs/ignore/0.4.17/ignore/struct.WalkBuilder.html#method.add_ignore) that contains your `**/*.lua` rule. Since it has low precedent, then I think other things would then be able to override it.

It's probably inconvenient for you to use this API since it requires a file path, but I would be open to adding a new API that accepted a `std::io::Read` implementation.

**Reply by JohnnyMorganz** [reply]
No worries, I figured out a way to do it after all (I didn't think hard enough originally!). I'm doing something similar to what you originally suggested, applying my default rule in the WalkBuilder, only if no glob patterns were passed on the command line, by using the `globset` crate. Thanks for your help
